scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileService.java

```
@@ -0,0 +1,152 @@
+                           DataFileRepository dataFileRepository,
+                           DatasetRepository datasetRepository) {
+        this.rootLocation = Paths.get(properties.getLocation());
+        this.dataFileRepository = dataFileRepository;
+        this.datasetRepository = datasetRepository;
+    }
+
+    /**
+     * Creates a file on the filesystem and, if successful, stores the DataFile
+     * data in the database. A Dataset has many DataFiles and a DataFile belongs to one Dataset.
+     *
+     * @param file        A MultipartFile.
+     * @param datasetId   The primary key of the Dataset.
+     * @param username    The user / owner of the Dataset.
+     * @param datasetName The name of the Dataset.
+     * @return boolean indicating success.
+     */
+    @Override
+    public boolean createFile(MultipartFile file, Long datasetId, String username, String datasetName) {
+        boolean success;
+        String fileName = file.getOriginalFilename();
+        String type = file.getContentType();
+
+        try {
+            success = createFileInDirectory(username, datasetName, file);
+        } catch (DataFileException e) {
+            throw new DataFileException("File could not be saved in the file system", e);
+        }
+
+        // Only store file in database if saving file on filesystem was successful
+        if (success) {
+            Optional<DataFile> dataFile = dataFileRepository.findByName(fileName);
+            if (dataFile.isPresent() &&
+                    dataFile.get().getDataset().getId().equals(datasetId) &&
+                    dataFile.get().getType().equals(type)) {
+                // if a file with the same name & type already exists in the dataset, update time stamp
+                Long dataFileId = dataFile.get().getId();
+                editFile(dataFileId);
+            } else {
+                Optional<Dataset> dataset = datasetRepository.findById(datasetId);
+                if (dataset.isPresent()) {
```

what if it isn't? The method will still return `true`, won't it? I would throw an exception if there is no such database, since that should not happen.